cpp/src/groupby/groupby.cu

```diff
@@ -48,32 +50,43 @@ std::pair<std::unique_ptr<table>, std::vector<aggregation_result>>
 groupby::dispatch_aggregation(std::vector<aggregation_request> const& requests,
                               cudaStream_t stream,
                               rmm::mr::device_memory_resource* mr) {
+  // If sort groupby has been called once on this groupby object, then
+  // always use sort groupby from now on. Because once keys are sorted, 
+  // all the aggs that can be done by hash groupby are efficiently done by
+  // sort groupby as well.
   // Only use hash groupby if the keys aren't sorted and all requests can be
   // satisfied with a hash implementation
   if (not _keys_are_sorted and
+      not _helper and
       detail::hash::can_use_hash_groupby(_keys, requests)) {
     return detail::hash::groupby(_keys, requests, _ignore_null_keys, stream,
                                  mr);
   } else {
-    return detail::sort::groupby(_keys, requests, stream, mr);
+    return sort_aggregate(requests, stream, mr);
   }
 }
 
+// Destructor
+// Needs to be in source file because sort_groupby_helper was forward declared
+groupby::~groupby() = default;
+
 namespace {
 /// Make an empty table with appropriate types for requested aggs
-auto empty_results(std::vector<aggregation_request> const& requests) {
+template <typename aggregation_request_t, typename F>
+auto templated_empty_results(std::vector<aggregation_request_t> const& requests, F get_kind) {
+  
   std::vector<aggregation_result> empty_results;
 
   std::transform(
       requests.begin(), requests.end(), std::back_inserter(empty_results),
-      [](auto const& request) {
+      [&get_kind](auto const& request) {
         std::vector<std::unique_ptr<column>> results;
 
         std::transform(
             request.aggregations.begin(), request.aggregations.end(),
-            std::back_inserter(results), [&request](auto const& agg) {
+            std::back_inserter(results), [&request, get_kind](auto const& agg) {
               return make_empty_column(experimental::detail::target_type(
-                  request.values.type(), agg->kind));
+                  request.values.type(), get_kind(agg)));
```

Comment:
If you decide to go with option 2 regarding aggregation objects, you wouldn't need to templatize this and `agg->kind` would work for both.

Reply:
I'm afraid I've gone with separate data-structures. It made the code downstream more succinct.